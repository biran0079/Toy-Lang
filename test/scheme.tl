fun cons(a,b){
  fun f(i) {
    if(i==0)  return a;
    return b;
  }
  return f;
}
fun null() {}
fun car(p){
  return p(0);  
}

fun cdr(p){
  return p(1);  
}

fun slistToList(l){
  if(l==null)  return [];
  return [car(l)] + slistToList(cdr(l));
}

fun range(n){
  res=null;
  while(n>0){
    res = cons(n, res);
    n=n-1;
  }
  return res;
}

fun map(f, l){
  if(l==null)return null;
  return cons(f(car(l)) , map(f,cdr(l)));
}
fun square(n){
  return n*n;  
}
fun filter(f, l){
  if(l==null) return null;
  if(f(car(l)))  return cons(car(l), filter(f, cdr(l)));
  return filter(f,cdr(l));
}
fun isOdd(n){
  return n%2==1;  
}
fun sieve(l){
  if(l==null)return null;
  i=car(l);
  return cons(i, sieve(filter(lambda(n){return n%i!=0;},cdr(l))));
}
fun flat(l){
  if(l==null) return null;
  return append(car(l), flat(cdr(l)));
}
fun append(l1, l2){
  if(l1==null)  return l2;
  return cons(car(l1), append(cdr(l1), l2));
}
print(slistToList(map(square, range(10))));
print(slistToList(filter(isOdd, range(10))));
print(slistToList(sieve(cdr(range(20)))));
print(slistToList(append(range(4), range(5))));
print(slistToList(flat(cons(range(4), cons(range(5), null)))));
fun perm(l){
  if(l==null)  return cons(null, null);
  return flat(
      map(lambda(x){
              return map(lambda(l){ return cons(x, l); },
                  perm(filter(lambda(y) {return y!=x;}, l)));
        }, l));
}

fun printSlistOfSlist(l){
  if(l==null) return null;
  print(slistToList(car(l)));
  return printSlistOfSlist(cdr(l));
}

printSlistOfSlist(perm(range(3)));
