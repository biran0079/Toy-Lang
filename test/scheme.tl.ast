(stmts (fun (id cons) (ids (id a) (id b)) (stmts (fun (id f) (ids (id i)) (stmts (if (== (id i) (int 0)) (return (id a))) (return (id b)))) (return (id f)))) (fun (id null) (ids) (stmts)) (fun (id car) (ids (id p)) (stmts (return (tail_call (id p) (exps (int 0)))))) (fun (id cdr) (ids (id p)) (stmts (return (tail_call (id p) (exps (int 1)))))) (fun (id slistToList) (ids (id l)) (stmts (if (== (id l) (id null)) (return (list))) (return (+ (list (call (id car) (exps (id l)))) (call (id slistToList) (exps (call (id cdr) (exps (id l))))))))) (fun (id range) (ids (id n)) (stmts (= (id res) (id null)) (while (> (id n) (int 0)) (stmts (= (id res) (call (id cons) (exps (id n) (id res)))) (= (id n) (- (id n) (int 1))))) (return (id res)))) (fun (id map) (ids (id f) (id l)) (stmts (if (== (id l) (id null)) (return (id null))) (return (tail_call (id cons) (exps (call (id f) (exps (call (id car) (exps (id l))))) (call (id map) (exps (id f) (call (id cdr) (exps (id l)))))))))) (fun (id square) (ids (id n)) (stmts (return (* (id n) (id n))))) (fun (id filter) (ids (id f) (id l)) (stmts (if (== (id l) (id null)) (return (id null))) (if (call (id f) (exps (call (id car) (exps (id l))))) (return (call (id cons) (exps (call (id car) (exps (id l))) (call (id filter) (exps (id f) (call (id cdr) (exps (id l))))))))) (return (tail_call (id filter) (exps (id f) (call (id cdr) (exps (id l)))))))) (fun (id isOdd) (ids (id n)) (stmts (return (== (% (id n) (int 2)) (int 1))))) (fun (id sieve) (ids (id l)) (stmts (if (== (id l) (id null)) (return (id null))) (= (id i) (call (id car) (exps (id l)))) (return (tail_call (id cons) (exps (id i) (call (id sieve) (exps (call (id filter) (exps (fun (id lambda) (ids (id n)) (stmts (return (!= (% (id n) (id i)) (int 0))))) (call (id cdr) (exps (id l)))))))))))) (fun (id flat) (ids (id l)) (stmts (if (== (id l) (id null)) (return (id null))) (return (tail_call (id append) (exps (call (id car) (exps (id l))) (call (id flat) (exps (call (id cdr) (exps (id l)))))))))) (fun (id append) (ids (id l1) (id l2)) (stmts (if (== (id l1) (id null)) (return (id l2))) (return (tail_call (id cons) (exps (call (id car) (exps (id l1))) (call (id append) (exps (call (id cdr) (exps (id l1))) (id l2)))))))) (call (id print) (exps (call (id slistToList) (exps (call (id map) (exps (id square) (call (id range) (exps (int 10))))))))) (call (id print) (exps (call (id slistToList) (exps (call (id filter) (exps (id isOdd) (call (id range) (exps (int 10))))))))) (call (id print) (exps (call (id slistToList) (exps (call (id sieve) (exps (call (id cdr) (exps (call (id range) (exps (int 20))))))))))) (call (id print) (exps (call (id slistToList) (exps (call (id append) (exps (call (id range) (exps (int 4))) (call (id range) (exps (int 5))))))))) (call (id print) (exps (call (id slistToList) (exps (call (id flat) (exps (call (id cons) (exps (call (id range) (exps (int 4))) (call (id cons) (exps (call (id range) (exps (int 5))) (id null))))))))))) (fun (id perm) (ids (id l)) (stmts (if (== (id l) (id null)) (return (call (id cons) (exps (id null) (id null))))) (return (tail_call (id flat) (exps (call (id map) (exps (fun (id lambda) (ids (id x)) (stmts (return (call (id map) (exps (fun (id lambda) (ids (id l)) (stmts (return (call (id cons) (exps (id x) (id l)))))) (call (id perm) (exps (call (id filter) (exps (fun (id lambda) (ids (id y)) (stmts (return (!= (id y) (id x))))) (id l)))))))))) (id l)))))))) (fun (id printSlistOfSlist) (ids (id l)) (stmts (if (== (id l) (id null)) (return (id null))) (call (id print) (exps (call (id slistToList) (exps (call (id car) (exps (id l))))))) (return (tail_call (id printSlistOfSlist) (exps (call (id cdr) (exps (id l)))))))) (call (id printSlistOfSlist) (exps (call (id perm) (exps (call (id range) (exps (int 3))))))))
