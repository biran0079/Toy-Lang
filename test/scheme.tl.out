[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
[1, 3, 5, 7, 9]
[2, 3, 5, 7, 11, 13, 17, 19]
[1, 2, 3, 4, 1, 2, 3, 4, 5]
[1, 2, 3, 4, 1, 2, 3, 4, 5]
[1, 2, 3]
[1, 3, 2]
[2, 1, 3]
[2, 3, 1]
[3, 1, 2]
[3, 2, 1]
